torch/csrc/jit/codegen/cuda/ir_internal_nodes.h

```diff
@@ -267,5 +267,168 @@ struct TORCH_CUDA_API Reorder : public Expr {
+  bool sameAs(const ForLoop* other) const;
+
+ private:
+  Int* const index_;
+  IterDomain* const range_;
+  std::vector<const Expr*> body_;
+};
+
+
+/*
+ * IfThenElse provides scoping for an boolean operator. Exprs placed in its body
+ * are considered inside the scope of the if statement. In the future the
+ * implementation should look quite different so that we can do proper
+ * dependency annalysis like in Fusion.
+ *
+ * TODO: Change implmentation of Exprs contained in the scope to be more similar
+ * to Fusion where we can do proper dependency analysis.
+ */
+struct TORCH_API IfThenElse : public Expr {
+  ~IfThenElse() = default;
+  IfThenElse(
+      Val* _cond,
+      const std::vector<const Expr*>& _if_body,
+      const std::vector<const Expr*>& _else_body = {});
+
+  IfThenElse(const IfThenElse& other) = delete;
+  IfThenElse& operator=(const IfThenElse& other) = delete;
+
+  IfThenElse(IfThenElse&& other) = delete;
+  IfThenElse& operator=(IfThenElse&& other) = delete;
+
+  Val* cond() const noexcept {
+    return cond_;
+  }
+  const std::vector<const Expr*>& if_body() const noexcept {
+    return if_body_;
+  }
+  const std::vector<const Expr*>& else_body() const noexcept {
+    return else_body_;
+  }
+
+  void add_if_expr(const Expr* e) {
+    if_body_.push_back(e);
+  }
+  void add_else_expr(const Expr* e) {
+    else_body_.push_back(e);
+  }
+
+  bool hasElse() const noexcept {
+    return !else_body_.empty();
+  }
+
+  bool sameAs(const IfThenElse* other) const;
+  
+ private:
+  // TODO: Why is the pointer const and not what's in the object?
+  Val* const cond_;
+  std::vector<const Expr*> if_body_;
+  std::vector<const Expr*> else_body_;
+};
+
+
+/*
+ * TODO: Fill out TensorIndex, which is a list of Ints used to directly index a
+ * TensorView. It is not the flattened index, which needs to be computed using
+ * stride information.
+ */
+struct TORCH_API TensorIndex : public Val {
+  ~TensorIndex() = default;
+
+  TensorIndex(const TensorIndex& other) = delete;
+  TensorIndex& operator=(const TensorIndex& other) = delete;
+
+  TensorIndex(TensorIndex&& other) = delete;
+  TensorIndex& operator=(TensorIndex&& other) = delete;
+
+  TensorIndex(std::vector<Int*> _indices)
+      : Val(ValType::TensorIndex), indices_(_indices) {}
+
+  std::vector<Int*>::size_type size() const {
+    return indices_.size();
+  }
+
+  bool sameAs(const TensorIndex* const other) const;
+  //i here is int, as we want to accept negative value and ::size_type can be a uint.
+  Int* axis(int i) const;
+
+ private:
+  std::vector<Int*> indices_;
+};
+
+/*
+ * Allocate is a lower level Node that describes a buffer of memory that
+ * is required as an intermediate within a kernel.  The extent is the expression
+ * of the size of the buffer that is generated from the TensorView that describes
+ * the output of an operation.
+ *
+ * TODO:
+ * 1.) Should extent_ be an Expr vs a Val?  The Val is currently used to print
+ *     the Expr of the size().
```

Comment:
It should be a value as we could allocate based on something simple like `new Int(5)` which doesn't have any expression associated with it.